### ISSUE TITLE
ENH-000097264  Add the functionality of adjusting column widths of Crowdsource manager application

### ISSUE THREAD

**Issue by allisonmuise** [Esri/crowdsource-manager#187]
#1264 in arcgis-portal-app-templates

Add the functionality of adjusting column widths of Crowdsource manager application.
Salesforce ID: ENH-000097264
Salesforce Submitter: Christine Ma
Salesforce Submit Date: 6/17/2016 9:45 AM
Salesforce Bug Type: Product
Salesforce Severity: Medium
Repro Data: (n/a)
Work Around: 1. Provide the appropriate/comparable step by step workflow to help illustrate the request. 
Following the steps in this documentation (create maps and create application from the group)
http://solutions.arcgis.com/local-government/help/crowdsource-manager/get-started/setup-map/
User would like to have the functionality to adjust the column widths of the Crowdsource manager application.
application example: http://www.arcgis.com/apps/CrowdsourceManager/index.html
2. What is the problem feature X is trying to solve? (i.e. What is the problem/limitation/design choice that is preventing the customer from moving forward?) 
The view/wide columns makes it extremely cumbersome to read as one has to scroll through all the columns to find the information needed.
3. What are the benefits or changes that would result from the enhancement? (i.e. How does making this change help the customer's workflow or business?) 
Allowing for adjustable column width will allow the user to easily see all the information at one glance instead of having to scroll over constantly to see all the attributes associated with the feature.  This will result in higher efficiency.

Product: (n/a)
Functional Category: ArcGIS Online Application Templates
Client Platform: (n/a)
Version Found: (n/a)
Planned Version Fixed: (n/a)
Comment: (n/a)

**Comment by allisonmuise** [comment]
The end goal is to make it easier to view content. I'm open to other options such as wrapping long text and auto-sizing columns based on field sizes.

I know we are using a lot of core components in this table. I only want to do this if it is possible without tearing apart what we have already.

**Comment by CTLocalGovTeam** [comment]
@allisonmuise,

For data-viewer we have used calcite table which internally uses simple HTML table. It has a limitation that it does not provide on-the-fly column resizing. Examples of column resizing on the web usually make use of complex data grid controls, which we initially moved away from, so the table and its rendering can be lean.

We have another different options to make content viewing easier as listed below.

> 1. The width of the column would be fixed & long text will be wrapped. Do we want a height limit here? Otherwise large pieces of text will lead to very large cell heights.

![wrap lines](https://cloud.githubusercontent.com/assets/2959186/20875775/64e22850-bae3-11e6-85c1-0e880b866b23.PNG)

> 2. The width of the column will be floating depending on the text. No text will be cut off.

![no wrap](https://cloud.githubusercontent.com/assets/2959186/20875806/9e7106ae-bae3-11e6-84d6-37f90b620481.PNG)

Please let us know your thoughts on this.

--Pratik Gohil

**Comment by allisonmuise** [comment]
@CTLocalGovTeam : Do we need to choose between 1 or 2? Option 1 would help with long pieces of text, and Option 2 would help efficiently display fields with short text. If we can't allow users to adjust the width of the columns, it would make more sense to combine the two options. 

My suggestion: 

The space required for a field should be controlled by the longest of these values:
- the length of the field alias
- field length (non-domain fields)
- longest domain description (domain fields)

Allow field width to shrink/grow to accommodate the longest of these values. If the required field width is more than 1/2 the available table space, wrap the text. To keep long text from taking over the table, let's wrap to a max of 3 lines. Users can click the record to view the rest of the text in the details panel.

What do you think?

**Comment by CTLocalGovTeam** [comment]
@allisonmuise,

We are analysing this enhancement & keep you updated.

--Pratik Gohil

**Comment by CTLocalGovTeam** [comment]
@allisonmuise,

After analysing your approach, which seems quite complex, I want to propose a simpler approach that is easy to implement. I find a couple of challenges in your approach, like, retrieving the length of any text in pixels. This depends on font, font-size that a browser has used or substituted. 

In my proposed approach, we set the table to have column width between 120 & 300 px (can be made configurable in default.js). So that,

1. Each column will have a minimum width of 120px
2. And a maximum width of 300px

Now, the browser does the calculations and adjusts data that fit within 120px; keeping the column to width no less than 120px, irrespective of available width on screen. For cells that require higher width, the browser sets the column width up to 300px and wraps text up to 3 lines. Any lengthier text will be truncated. 

![image](https://cloud.githubusercontent.com/assets/2959186/21048946/f0215e2a-be37-11e6-954c-e4af88af8b1f.png)

Please let me know if this approach works for you.

--Pratik Gohil

**Comment by allisonmuise** [comment]
@CTLocalGovTeam : This is getting way more complicated that I was hoping. I'm going to push this to a release later in the year. We can re-assess then to see if there's any changes in the tech that will help make this easier.

**Comment by allisonmuise** [comment]
closing this. The new UI makes reviewing and editing attributes much easier.

**Comment by 3vivekb** [comment]
This is still a problem and has pretty inefficient table sizes:

![image](https://user-images.githubusercontent.com/4524205/33583095-735a225e-d90d-11e7-84c7-a6fbb40f360b.png)

The Styling should auto fit to the largest element in the table.

**Comment by allisonmuise** [comment]
@3vivekb I agree that the table is not a very efficient view of the data, but if you click a row in the table you get a much more efficient and read-able view in the Details panel. 

We did play around with adjusting the fields in far more conversations than are documented above and, in the end, making the table more responsive to the size of the data would have been a large development effort with very little payoff. Yes, a bit of real estate would have been gained, but calculating the field size required for number fields and large text fields dramatically slowed app loading speeds, and it didn't actually make that much difference in the read-ability of the data in the table. It looked a lot busier, and data was still getting cut off therefore you still needed to rely on the Details view to get a full picture of the record. 

For these reasons, we decided to enhance the Details display rather than pursue more flexible column widths. The table is still there for sorting, filtering, and otherwise identifying the records you'd like to work with, but it's not intended to be the optimal way to review the attributes of a feature - just as there are both table and popup views of attributes in a web map.
